### DISCUSSION TITLE
Detect db role switch (primary/secondary/replica)

### DISCUSSION THREAD

**Discussion by brihi4k** [sqlalchemy/sqlalchemy#6951] [Usage Questions] [answered]
I'm connecting to a postgres database that has a primary node and secondary node (called the replica). The primary node supports read/write queries and the replica supports read only queries. I create connection pools to both nodes and direct read only queries to the replica. Sometimes the nodes switch roles, primary becomes replica and replica becomes primary. This is done on the db side and using dns to swap the names/ips.

After the switch occurs the connections remain but are now pointing to the wrong node. I need to drop all connections and recreate them so they point to the correct node, but I don't know how to detect that the role switch occurred. Read/write queries fail with sqlalchemy.exc.InternalError: (psycopg2.errors.ReadOnlySqlTransaction), which I could use but it feels hacky/flakey.

Is there a more direct way to detect when the switches happen?

sqlachemy 1.3
postgres 12

**Comment by zzzeek** [comment, marked as answer]
well quickest way is to not use connection pooling, if you are using pgbouncer, then you can use NullPool and not have the problem that way.

another approach is to use an "optimistic" approach meaning you have to wait for something to fail, look at the error message to see that it's this condition, then do the .dispose() on your engines.    you can assemble event handlers into the engines to intercept all errors using the [handle_error](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=handle_error#sqlalchemy.events.ConnectionEvents.handle_error) event hook.   what you want to do is when the condition is detected, set the [.is_disconnect](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.ExceptionContext.is_disconnect) flag on the ExceptionContext to True, which should cause the pool to be invalidated.

it might also be possible to use a connection pool "on checkout" listener, or an engine-level "engine_connect" listener, similar to the "pre-ping" feature, where you test the connection that it's in the role you expect, and if not, do a reconnect.   you would adapt the recipe at https://docs.sqlalchemy.org/en/14/core/pooling.html#custom-legacy-pessimistic-ping to this case, where youre not only "pinging" the database, but also selecting the PG environment variables that indicate if its the read-only node or not and compare that to what's expected.  i would probably modify that recipe to do a connection.invalidate() that would cause the pooled connection to be re-connected.

I think what you want to do is very possible but it might be a little bit difficult to get it to work.  the above two recipes are how to do it but even when i have to deal with these, it tends to be kind of tricky to get it to do the right thing.  failure modes are hard.

**Reply by brihi4k** [reply]
would an engine.pool.dispose() accomplish the same thing as engine.dispose()?

**Reply by zzzeek** [reply]
pool.dispose() will close out all connections that are currently checked in, but for connections that are checked out, they go back to that pool and would not have any indication they should be invalidated.  The pool has a pool._invalidate() method that will mark a timestamp under which all connections older than that will be reconnected, though this is private API right now and you need to send a connection from the pool to it.

**Reply by brihi4k** [reply]
thanks!
